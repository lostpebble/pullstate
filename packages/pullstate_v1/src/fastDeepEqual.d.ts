declare module "fast-deep-equal/es6" {
  const equal: (a: any, b: any) => boolean;
  export = equal;
}
