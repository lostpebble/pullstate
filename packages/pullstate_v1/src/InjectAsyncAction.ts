import React from "react";
import {
  IAsyncActionBeckonOptions,
  IAsyncActionWatchOptions,
  IOCreateAsyncActionOutput,
  TPullstateAsyncBeckonResponse,
  TPullstateAsyncWatchResponse
} from "./async-types";
import { IPullstateAllStores } from "./PullstateCore";

export enum EAsyncActionInjectType {
  WATCH = "watch",
  BECKON = "beckon",
}

interface IPropsInjectAsyncActionBase<A, R, T extends string, N> {
  action: IOCreateAsyncActionOutput<A, R, T, N>;
  args?: A;
}

export interface IPropsInjectAsyncActionBeckon<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores>
  extends IPropsInjectAsyncActionBase<A, R, T, N> {
  type: EAsyncActionInjectType.BECKON;
  options?: IAsyncActionBeckonOptions<A, R, T, N, S>;
  children: (response: TPullstateAsyncBeckonResponse<R, T>) => React.ReactElement;
}

export interface IPropsInjectAsyncActionWatch<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores>
  extends IPropsInjectAsyncActionBase<A, R, T, N> {
  type: EAsyncActionInjectType.WATCH;
  children: (response: TPullstateAsyncWatchResponse<R, T, N>) => React.ReactElement;
  options?: IAsyncActionWatchOptions<A, R, T, N, S>;
}

export type TInjectAsyncActionProps = IPropsInjectAsyncActionBeckon | IPropsInjectAsyncActionWatch;

export function InjectAsyncAction(
  props: TInjectAsyncActionProps
): React.ReactElement {
  if (props.type === EAsyncActionInjectType.BECKON) {
    const response = props.action.useBeckon(props.args, props.options);
    return props.children(response);
  }

  const response = props.action.useWatch(props.args, props.options);
  return props.children(response);
}
