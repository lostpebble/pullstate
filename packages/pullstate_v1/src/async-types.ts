import { IPullstateAllStores } from "./PullstateCore";
import { TUpdateFunction } from "./Store";

type TPullstateAsyncUpdateListener = () => void;

// [ started, finished, result, updating, timeCached ]
export type TPullstateAsyncWatchResponse<R = any, T extends string = string, N = any> = [
  boolean,
  boolean,
  TAsyncActionResult<R, T, N>,
  boolean,
  number
];

// export type TPullstateAsync

// [ started, finished, result, updating, postActionResult ]
// export type TPullstateAsyncResponseCacheFull<R, T extends string> = [
//   boolean,
//   boolean,
//   TAsyncActionResult<R, T>,
//   boolean,
//   TAsyncActionResult<R, T> | true | null
// ];

// [finished, result, updating]
export type TPullstateAsyncBeckonResponse<R = any, T extends string = string, N = any> = [
  boolean,
  TAsyncActionResult<R, T, N>,
  boolean
];
// [result]
export type TPullstateAsyncRunResponse<R = any, T extends string = string, N = any> = Promise<TAsyncActionResult<R, T, N>>;

export interface IPullstateAsyncResultState {
  [key: string]: TPullstateAsyncWatchResponse<any, string, any>;
}

export interface IPullstateAsyncActionOrdState {
  [key: string]: number;
}

export enum EAsyncEndTags {
  THREW_ERROR = "THREW_ERROR",
  RETURNED_ERROR = "RETURNED_ERROR",
  UNFINISHED = "UNFINISHED",
  DORMANT = "DORMANT",
}

interface IAsyncActionResultBase<T extends string> {
  message: string;
  tags: (EAsyncEndTags | T)[];
}

export interface IAsyncActionResultPositive<R, T extends string> extends IAsyncActionResultBase<T> {
  error: false;
  payload: R;
  errorPayload: null;
}

export interface IAsyncActionResultNegative<T extends string, N = unknown> extends IAsyncActionResultBase<T> {
  error: true;
  errorPayload: N;
  payload: null;
}

export type TAsyncActionResult<R, T extends string, N> =
  IAsyncActionResultPositive<R, T>
  | IAsyncActionResultNegative<T, N>;

// Order of new hook functions:

// shortCircuitHook = ({ args, stores }) => cachable response | false     - happens only on uncached action
// cacheBreakHook = ({ args, stores, result }) => true | false            - happens only on cached action
// postActionHook = ({ args, result, stores }) => void | new result       - happens on all actions, after the async / short circuit has resolved
// ----> postActionHook potentially needs a mechanism which allows it to run only once per new key change (another layer caching of some sorts expiring on key change)

export type TPullstateAsyncShortCircuitHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  stores: S;
}) => TAsyncActionResult<R, T, N> | false;

export type TPullstateAsyncCacheBreakHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  result: TAsyncActionResult<R, T, N>;
  stores: S;
  timeCached: number;
}) => boolean;

export enum EPostActionContext {
  WATCH_HIT_CACHE = "WATCH_HIT_CACHE",
  BECKON_HIT_CACHE = "BECKON_HIT_CACHE",
  RUN_HIT_CACHE = "RUN_HIT_CACHE",
  READ_HIT_CACHE = "READ_HIT_CACHE",
  READ_RUN = "READ_RUN",
  SHORT_CIRCUIT = "SHORT_CIRCUIT",
  DIRECT_RUN = "DIRECT_RUN",
  BECKON_RUN = "BECKON_RUN",
  CACHE_UPDATE = "CACHE_UPDATE",
}

export type TPullstateAsyncPostActionHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  result: TAsyncActionResult<R, T, N>;
  stores: S;
  context: EPostActionContext;
}) => void;

export interface IAsyncActionReadOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  postActionEnabled?: boolean;
  cacheBreakEnabled?: boolean;
  key?: string;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>
}

export interface IAsyncActionBeckonOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionReadOptions<A, R, T, N, S> {
  ssr?: boolean;
  holdPrevious?: boolean;
  dormant?: boolean;
}

export interface IAsyncActionWatchOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionBeckonOptions<A, R, T, N, S> {
  initiate?: boolean;
}

export interface IAsyncActionUseOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionWatchOptions<A, R, T, N, S> {
  onSuccess?: (result: R, args: A) => void;
}

export interface IAsyncActionUseDeferOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends Omit<IAsyncActionReadOptions<A, R, T, N, S>, "key"> {
  key?: string;
  holdPrevious?: boolean;
  onSuccess?: (result: R, args: A) => void;
  clearOnSuccess?: boolean;
}

export interface IAsyncActionRunOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  treatAsUpdate?: boolean;
  ignoreShortCircuit?: boolean;
  respectCache?: boolean;
  key?: string;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>
  _asyncCache?: IPullstateAsyncCache;
  _stores?: S;
  _customContext?: any;
}

export interface IAsyncActionGetCachedOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  checkCacheBreak?: boolean;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>;
  key?: string;
}

export interface IGetCachedResponse<R, T extends string, N = any> {
  started: boolean;
  finished: boolean;
  result: TAsyncActionResult<R, T, N>;
  updating: boolean;
  existed: boolean;
  cacheBreakable: boolean;
  timeCached: number;
}

export interface IAsyncClearCacheOptions {
  notify?: boolean;
}

export interface IAsyncActionSetOrClearCachedValueOptions extends IAsyncClearCacheOptions {
  key?: string;
}

export interface IAsyncActionUpdateCachedOptions extends IAsyncActionSetOrClearCachedValueOptions {
  resetTimeCached?: boolean;
  runPostActionHook?: boolean;
}

export type TAsyncActionUse<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionUseOptions<A, R, T, N, S>
) => TUseResponse<A, R, T, N, S>;

export type TAsyncActionUseDefer<A, R, T extends string, N, S extends IPullstateAllStores> = (
  options?: IAsyncActionUseDeferOptions<A, R, T, N, S>
) => TUseDeferResponse<A, R, T, N, S>;

export type TAsyncActionBeckon<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionBeckonOptions<A, R, T, N, S>
) => TPullstateAsyncBeckonResponse<R, T, N>;

export type TAsyncActionWatch<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionWatchOptions<A, R, T, N, S>
) => TPullstateAsyncWatchResponse<R, T, N>;

export type TAsyncActionRun<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionRunOptions<A, R, T, N, S>
) => TPullstateAsyncRunResponse<R, T, N>;

export type TAsyncActionClearCache<A> = (args?: A, options?: IAsyncActionSetOrClearCachedValueOptions) => void;

export type TAsyncActionClearAllCache = (options?: IAsyncClearCacheOptions) => void;

export type TAsyncActionClearAllUnwatchedCache = (options?: IAsyncClearCacheOptions) => void;

export type TAsyncActionGetCached<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionGetCachedOptions<A, R, T, N, S>
) => IGetCachedResponse<R, T, N>;

export type TAsyncActionSetCached<A, R, T extends string, N> = (
  args: A,
  result: TAsyncActionResult<R, T, N>,
  options?: IAsyncActionSetOrClearCachedValueOptions
) => void;

export type TAsyncActionSetCachedPayload<A, R> = (args: A, payload: R, options?: IAsyncActionSetOrClearCachedValueOptions) => void;

export type TAsyncActionUpdateCached<A, R> = (
  args: A,
  updater: TUpdateFunction<R>,
  options?: IAsyncActionUpdateCachedOptions
) => void;
export type TAsyncActionRead<A, R, T extends string, N, S extends IPullstateAllStores> = (args?: A, options?: IAsyncActionReadOptions<A, R, T, N, S>) => R;

export type TAsyncActionDelayedRun<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args: A,
  options: IAsyncActionRunOptions<A, R, T, N, S> & { delay: number; clearOldRun?: boolean; immediateIfCached?: boolean }
) => () => void;

export interface IOCreateAsyncActionOutput<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> {
  use: TAsyncActionUse<A, R, T, N, S>;
  useDefer: TAsyncActionUseDefer<A, R, T, N, S>;
  read: TAsyncActionRead<A, R, T, N, S>;
  useBeckon: TAsyncActionBeckon<A, R, T, N, S>;
  useWatch: TAsyncActionWatch<A, R, T, N, S>;
  run: TAsyncActionRun<A, R, T, N, S>;
  delayedRun: TAsyncActionDelayedRun<A, R, T, N, S>;
  getCached: TAsyncActionGetCached<A, R, T, N, S>;
  setCached: TAsyncActionSetCached<A, R, T, N>;
  setCachedPayload: TAsyncActionSetCachedPayload<A, R>;
  updateCached: TAsyncActionUpdateCached<A, R>;
  clearCache: TAsyncActionClearCache<A>;
  clearAllCache: TAsyncActionClearAllCache;
  clearAllUnwatchedCache: TAsyncActionClearAllUnwatchedCache;
}

export interface IPullstateAsyncCache {
  results: IPullstateAsyncResultState;
  listeners: {
    [key: string]: {
      [watchId: string]: TPullstateAsyncUpdateListener;
    };
  };
  actions: {
    [key: string]: () => Promise<TAsyncActionResult<any, string, any>>;
  };
  actionOrd: IPullstateAsyncActionOrdState;
}

export type TPullstateAsyncAction<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args: A,
  stores: S,
  customContext: any
) => Promise<TAsyncActionResult<R, T, N>>;

export interface ICreateAsyncActionOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  forceContext?: boolean;
  // clientStores?: S;
  shortCircuitHook?: TPullstateAsyncShortCircuitHook<A, R, T, N, S>;
  cacheBreakHook?: TPullstateAsyncCacheBreakHook<A, R, T, N, S>;
  postActionHook?: TPullstateAsyncPostActionHook<A, R, T, N, S>;
  subsetKey?: (args: A) => any;
  actionId?: string | number;
}

// action.use() types

export interface IUseDebouncedExecutionOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  validInput?: (args: A) => boolean;
  equality?: ((argsPrev: A, argsNew: A) => boolean) | any;
  executeOptions?: Omit<IAsyncActionRunOptions<A, R, T, N, S>, "key" | "cacheBreak">;
  watchLastValid?: boolean;
}

export type TRunWithPayload<R> = (func: (payload: R) => any) => any;

export interface IBaseObjResponseUse<A, R, T extends string, N, S extends IPullstateAllStores> {
  execute: (runOptions?: IAsyncActionRunOptions<A, R, T, N, S>) => TPullstateAsyncRunResponse<R, T, N>;
}

export interface IBaseObjResponseUseDefer<A, R, T extends string, N, S extends IPullstateAllStores> {
  execute: (args?: A, runOptions?: Omit<IAsyncActionRunOptions<A, R, T, N, S>, "key" | "cacheBreak">) => TPullstateAsyncRunResponse<R, T, N>;
  hasCached: (args?: A, options?: { successOnly?: boolean } & Omit<IAsyncActionGetCachedOptions<A, R, T, N, S>, "key">) => boolean;
  unwatchExecuted: () => void;
  useDebouncedExecution: (args: A, delay: number, options?: IUseDebouncedExecutionOptions<A, R, T, N, S>) => void;
  args: A;
  key: string;
}

export interface IBaseObjResponse<R, T extends string, N> {
  isLoading: boolean;
  isFinished: boolean;
  isUpdating: boolean;
  isStarted: boolean;
  // isSuccess: boolean;
  // isFailure: boolean;
  clearCached: () => void;
  updateCached: (updater: TUpdateFunction<R>, options?: IAsyncActionUpdateCachedOptions) => void;
  setCached: (result: TAsyncActionResult<R, T, N>, options?: IAsyncActionSetOrClearCachedValueOptions) => void;
  setCachedPayload: (payload: R, options?: IAsyncActionSetOrClearCachedValueOptions) => void;
  endTags: (T | EAsyncEndTags)[];
  renderPayload: TRunWithPayload<R>;
  message: string;
  raw: TPullstateAsyncWatchResponse<R, T, N>;
}

export interface IBaseObjSuccessResponse<R, T extends string, N> extends IBaseObjResponse<R, T, N> {
  payload: R;
  errorPayload: null;
  error: false;
  isSuccess: true;
  isFailure: false;
}

export interface IBaseObjErrorResponse<R, T extends string, N> extends IBaseObjResponse<R, T, N> {
  payload: null;
  errorPayload: N;
  error: true;
  isFailure: true;
  isSuccess: false;
}

export type TUseResponse<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> =
  (IBaseObjSuccessResponse<R, T, N>
    | IBaseObjErrorResponse<R, T, N>) & IBaseObjResponseUse<A, R, T, N, S>;

export type TUseDeferResponse<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> =
  (IBaseObjSuccessResponse<R, T, N>
    | IBaseObjErrorResponse<R, T, N>) & IBaseObjResponseUseDefer<A, R, T, N, S>;
