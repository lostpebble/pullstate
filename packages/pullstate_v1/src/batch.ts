import { globalClientState } from "./globalClientState";

interface IBatchState {
  uiBatchFunction: ((updates: () => void) => void);
}

const batchState: Partial<IBatchState> = {};

export function setupBatch({ uiBatchFunction }: IBatchState) {
  batchState.uiBatchFunction = uiBatchFunction;
}

export function batch(runUpdates: () => void) {
  if (globalClientState.batching) {
    throw new Error("Pullstate: Can't enact two batch() update functions at the same time-\n" +
      "make sure you are not running a batch() inside of a batch() by mistake.");
  }

  globalClientState.batching = true;

  try {
    runUpdates();
  } finally {
    if (batchState.uiBatchFunction) {
      batchState.uiBatchFunction(() => {
        Object.values(globalClientState.flushStores).forEach(store => store.flushBatch(true));
      });
    } else {
      Object.values(globalClientState.flushStores).forEach(store => store.flushBatch(true));
    }
    globalClientState.flushStores = {};
    globalClientState.batching = false;
  }
}
