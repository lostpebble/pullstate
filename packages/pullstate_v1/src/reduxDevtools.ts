import { IPullstateAllStores } from "./PullstateCore";

interface IORegisterInDevtoolsOptions {
  namespace?: string;
}

export function registerInDevtools(stores: IPullstateAllStores, { namespace = "" }: IORegisterInDevtoolsOptions = {}) {
  const devToolsExtension = typeof window !== "undefined" ? (window as any)?.__REDUX_DEVTOOLS_EXTENSION__ : undefined;

  if (devToolsExtension) {
    for (const key of Object.keys(stores)) {
      const store = stores[key];

      const devTools = devToolsExtension.connect({ name: `${namespace}${key}` });
      devTools.init(store.getRawState());
      let ignoreNext = false;
      /*store.subscribe(
          (state) => {
            if (ignoreNext) {
              ignoreNext = false;
              return;
            }
            devTools.send("Change", state);
          },
          () => {}
        );*/
      store.subscribe(
        (s) => s,
        (watched) => {
          if (ignoreNext) {
            ignoreNext = false;
            return;
          }
          devTools.send("Change", watched);
        }
      );

      devTools.subscribe((message: { type: string; state: any }) => {
        if (message.type === "DISPATCH" && message.state) {
          ignoreNext = true;
          const parsed = JSON.parse(message.state);
          store.replace(parsed);
        }
      });
    }
  }
}
