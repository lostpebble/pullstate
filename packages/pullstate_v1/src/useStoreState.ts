import isEqual from "fast-deep-equal/es6";
// S = State
// SS = Sub-state
import { useEffect, useRef, useState } from "react";
import { Store } from "./Store";

// const isEqual = require("fast-deep-equal/es6");

export interface IUpdateRef {
  shouldUpdate: boolean;
  onStoreUpdate: (() => void) | null;
  getSubState?: (state: any) => any;
  currentSubState: any;
  setInitial: boolean;
}

export interface IUpdateRefNew {
  state: any;
  initialized: boolean;
}

const onServer = typeof window === "undefined";

function useStoreState<S extends object = any>(store: Store<S>): S;
function useStoreState<S extends object = any, SS = any>(
  store: Store<S>,
  getSubState: (state: S) => SS,
  deps?: ReadonlyArray<any>
): SS;
function useStoreState(store: Store, getSubState?: (state: any) => any, deps?: ReadonlyArray<any>): any {
  const updateRef = useRef<IUpdateRefNew>({ state: undefined, initialized: false });

  if (!updateRef.current.initialized) {
    updateRef.current.state = getSubState ? getSubState(store.getRawState()) : store.getRawState();
    updateRef.current.initialized = true;
  }

  // useState with only a simple value to prevent double equality checks for the state
  const [, setUpdateTrigger] = useState(0);

  // const [current, setCurrent] = useState(() => {
  //   return getSubState ? getSubState(store.getRawState()) : store.getRawState();
  // });
  // updateRef.current.state = current;

  useEffect(() => {
    const effectState = { shouldUpdate: true };

    function update() {
      if (effectState.shouldUpdate) {
        const nextSubState = getSubState
          ? getSubState(store.getRawState())
          : store.getRawState();

        if (!isEqual(updateRef.current.state, nextSubState)) {
          // final check again before actually running state update (might prevent no-op errors with React)
          if (effectState.shouldUpdate) {
            updateRef.current.state = nextSubState;
            setUpdateTrigger((val) => val + 1);
          }
        }
      }
    }

    store._addUpdateListener(update);
    // This ensures we get the latest updated value, even if it has changed
    // between the initial rendering and the effect
    update();

    return () => {
      effectState.shouldUpdate = false;
      store._removeUpdateListener(update);
    };
  }, deps ?? []);

  if (deps !== undefined) {
    const prevDeps = useRef<ReadonlyArray<any>>(deps);
    if (!isEqual(deps, prevDeps)) {
      updateRef.current.state = getSubState!(store.getRawState());
    }
  }

  return updateRef.current.state;
}

/*
function useStoreState(store: Store, getSubState?: (state: any) => any, deps?: ReadonlyArray<any>): any {
  const updateRef = useRef<IUpdateRef>({
    shouldUpdate: true,
    onStoreUpdate: null,
    getSubState,
    currentSubState: null,
    setInitial: false,
  });

  const [, setUpdateTrigger] = useState(0);

  if (!updateRef.current.setInitial) {
    updateRef.current.currentSubState = updateRef.current.getSubState
      ? updateRef.current.getSubState(store.getRawState())
      : store.getRawState();
    updateRef.current.setInitial = true;
  }

  if (updateRef.current.onStoreUpdate === null) {
    updateRef.current.onStoreUpdate = function onStoreUpdate() {
      const nextSubState = updateRef.current.getSubState
        ? updateRef.current.getSubState(store.getRawState())
        : store.getRawState();
      if (updateRef.current.shouldUpdate && !isEqual(updateRef.current.currentSubState, nextSubState)) {
        // final check again before actually running state update (might prevent no-op errors with React)
        if (updateRef.current.shouldUpdate) {
          updateRef.current.currentSubState = nextSubState;
          setUpdateTrigger((val) => val + 1);
        }
      }
    };

    if (!onServer) {
      store._addUpdateListener(updateRef.current.onStoreUpdate!);
    }
  }

  useEffect(() => {
    updateRef.current.shouldUpdate = true;

    return () => {
      updateRef.current.shouldUpdate = false;
      store._removeUpdateListener(updateRef.current.onStoreUpdate!);
    };
  }, []);

  if (deps !== undefined) {
    const prevDeps = useRef<ReadonlyArray<any>>(deps);
    if (!isEqual(deps, prevDeps)) {
      updateRef.current.getSubState = getSubState;
      updateRef.current.currentSubState = getSubState!(store.getRawState());
    }
  }

  console.log(updateRef.current);

  return updateRef.current.currentSubState;
}
*/

export { useStoreState };
