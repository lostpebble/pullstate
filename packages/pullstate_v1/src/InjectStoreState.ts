import React from "react";
import { Store } from "./Store";
import { useStoreState } from "./useStoreState";

export interface IPropsInjectStoreState<S extends object = any, SS extends any = any> {
  store: Store<S>;
  on?: (state: S) => SS;
  children: (output: SS) => React.ReactElement;
}

export function InjectStoreState<S extends object = any, SS = any>({
  store,
  on = s => s as any,
  children,
}: IPropsInjectStoreState<S, SS>): React.ReactElement {
  const state: SS = useStoreState(store, on);
  return children(state);
}
