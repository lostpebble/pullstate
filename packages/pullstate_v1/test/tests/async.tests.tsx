import { useStoreState } from "../../src/useStoreState";
import React from "react";
import ReactDOMServer from "react-dom/server";
import { PullstateProvider, Store } from "../../src";
import { createTestBasics, IOGetUserInput, IUserStore } from "./TestSetup";
import { IOCreateAsyncActionOutput } from "../../src/async-types";

const beautifyHtml = require("js-beautify").html;

interface ITestProps {
  UserStore: Store<IUserStore>;
  ChangeToNewUserAsyncAction: IOCreateAsyncActionOutput<IOGetUserInput>;
}

const UninitiatedUserAction = ({ UserStore, ChangeToNewUserAsyncAction }: ITestProps) => {
  // const [userId, setUserId] = useState(0);
  const { user, userId } = useStoreState(UserStore, s => ({ user: s.user, userId: s.currentUserId }));
  const [started, finished, result, updating] = ChangeToNewUserAsyncAction.useWatch({ userId: 1 });

  return (
    <div>
      <span>
        {started ? (finished ? `Got new user` : `Getting new user`) : `Haven't initiated getting new user`}
      </span>
      {user !== null && (
        <div id="user-box">
          <h2>Hello, {user.name}</h2>
          <h3>aka: {user.userName}</h3>
        </div>
      )}
      {!started && (
        <button
          id="uninitiated-get-user-button"
          onClick={() => ChangeToNewUserAsyncAction.run({}, { treatAsUpdate: true })}>
          Initiate Get User for ID: {userId}
        </button>
      )}
    </div>
  );
};

const InitiatedNextUser = ({ UserStore, ChangeToNewUserAsyncAction }: ITestProps) => {
  const user = useStoreState(UserStore, s => s.user);
  const [finished] = ChangeToNewUserAsyncAction.useBeckon({ userId: 1 });

  return (
    <div>
      <span>{finished ? `User loaded` : `Loading user`}</span>
      {user !== null && (
        <div id="user-box">
          <h2>Hello, {user.name}</h2>
          <h3>aka: {user.userName}</h3>
        </div>
      )}
      <button
        disabled={!finished}
        id="initiated-get-user-button"
        onClick={() => ChangeToNewUserAsyncAction.run({})}>
        Get next user
      </button>
    </div>
  );
};

const App = (props: ITestProps) => {
  return (
    <div>
      <h1>Async Test</h1>
      <InitiatedNextUser {...props} />
      <UninitiatedUserAction {...props} />
    </div>
  );
};

describe("Async rendering", () => {
  it("renders our initial state without pre-resolved async", () => {
    const { ChangeToNewUserAsyncAction, UserStore } = createTestBasics();

    const reactHtml = ReactDOMServer.renderToString(<App {...{ ChangeToNewUserAsyncAction, UserStore }} />);
    expect(beautifyHtml(reactHtml)).toMatchSnapshot();
  });

  it("renders our initial state with some pre-resolved async state", async () => {
    const { ChangeToNewUserAsyncAction, UserStore, PullstateCore } = createTestBasics();

    const instance = PullstateCore.instantiate({ ssr: false });
    await instance.runAsyncAction(ChangeToNewUserAsyncAction, { userId: 1 });

    const reactHtml = ReactDOMServer.renderToString(
      <PullstateProvider instance={instance}>
        <App {...{ ChangeToNewUserAsyncAction, UserStore }} />
      </PullstateProvider>
    );
    expect(beautifyHtml(reactHtml)).toMatchSnapshot();
  });
});
