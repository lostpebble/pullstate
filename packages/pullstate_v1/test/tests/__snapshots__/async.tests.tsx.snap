// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async rendering renders our initial state with some pre-resolved async state 1`] = `
"<div>
    <h1>Async Test</h1>
    <div><span>User loaded</span>
        <div id=\\"user-box\\">
            <h2>Hello,
                <!-- -->Dave</h2>
            <h3>aka:
                <!-- -->davej</h3>
        </div><button id=\\"initiated-get-user-button\\">Get next user</button>
    </div>
    <div><span>Got new user</span>
        <div id=\\"user-box\\">
            <h2>Hello,
                <!-- -->Dave</h2>
            <h3>aka:
                <!-- -->davej</h3>
        </div>
    </div>
</div>"
`;

exports[`Async rendering renders our initial state without pre-resolved async 1`] = `
"<div data-reactroot=\\"\\">
    <h1>Async Test</h1>
    <div><span>Loading user</span><button disabled=\\"\\" id=\\"initiated-get-user-button\\">Get next user</button></div>
    <div><span>Getting new user</span></div>
</div>"
`;
